lib/commands.coffee

```diff
@@ -51,12 +52,26 @@ class Commands
       atom.notifications.addError error.message, {detail: error.stack, dismissable: true}
 
   nextError: ->
-    if not @messages or (next = @messages.next()).done
-      next = (@messages = @linter.views.getMessages().values()).next()
-    return if next.done # There's no errors
-    message = next.value
-    return unless message.filePath
-    return unless message.range
+    if @index?
+      @index++
+    else
+      @index = 0
```

Comment:
Like the order problem this will probably require a more structured message list to operate efficiently.  Something that knows which file something is in and when inserted does an ordered insert based on error start position.